test/test.sh

```diff
@@ -0,0 +1,297 @@
+#!/bin/bash -ue
+#
+# Copyright 2012-2013 OCamlPro
+#
+# All rights reserved.This file is distributed under the terms of the
+# GNU Lesser General Public License version 2.1 with linking
+# exception.
+#
+# TypeRex is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# Lesser GNU General Public License for more details.
+#
+
+shopt -s nullglob
+
+ROOT=$(git rev-parse --show-toplevel)
+OCAMLFORMAT="$ROOT"/_build/dbg/src/ocamlformat.exe
+cd $ROOT/test
+
+UPDATE=
+GIT=
+SHOW=
+SHOWCMD=
+HTML=
+
+usegit() {
+    printf "%-12s\t\e[34mgit %s\e[m\n" "" "$*";
+    git "$@";
+}
+
+is_file_on_git() {
+    [ $# -eq 1 ]; f=$1
+    git ls-files $f --error-unmatch >/dev/null 2>&1
+}
+
+while [ $# -gt 0 ]; do
+    case "$1" in
+        --update|-u)
+            UPDATE=1
+            ;;
+        --git-update)
+            if ! git diff --exit-code -- . >/dev/null; then
+                echo -e "\e[1mWarning:\e[m unstaged changes in test/"
+                echo "You may want to do 'git checkout -- test/' or"\
+                     "'git add -u -- test/' first."
+                exit 1
+            fi
+            UPDATE=1
+            GIT="usegit "
+            HTML=1
+            ;;
+        --ocamlformat)
+            if [ $# -le 1 ]; then echo "Error: $1 needs an argument"; exit 1; fi
+            shift;
+            OCAMLFORMAT=$1
+            ;;
+        --show)
+            SHOW=1
+            ;;
+        --meld)
+            SHOW=1
+            SHOWCMD="meld"
+            ;;
+        --html)
+            HTML=1
+            ;;
+        *)
+            cat <<EOF >/dev/stderr
+Usage:
+  -u --update          update the files according to the current results
+  --git-update         update the files and state the changes in git
+  --ocamlformat <prg>  use this ocamlformat exe
+  --show               show a diff of changed results
+  --meld               show progressions/regressions using meld
+  --html               generate an html page showing the diff of failing tests
+EOF
+            exit 1
+    esac
+    shift
+done
+
+TMP=$(mktemp -d /tmp/ocamlformat-test.XXXXX)
+trap "rm -rf /tmp/ocamlformat-${TMP#/tmp/ocamlformat-}" EXIT
+
+ocamlformat() {
+    [ $# -eq 1 ]
+    opts=$(cat $1.opts 2>/dev/null || true)
+    tmpfile=$TMP/$(basename $1)
+    "$OCAMLFORMAT" $opts "$1" >$tmpfile 2>&1 || echo "NON-ZERO EXIT CODE" > $tmpfile
```

Comment:
Note that the `--output` option could be used instead of redirecting stdout for slight perf gain since ocamlformat will generate a temp file when outputting to stdout. But I guess that it is desirable to redirect all output in case of errors.